k4FWCore/include/k4FWCore/FunctionalUtils.h

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c) 2014-2024 Key4hep-Project.
+ *
+ * This file is part of Key4hep.
+ * See https://key4hep.github.io/key4hep-doc/ for further info.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef FWCORE_FUNCTIONALUTILS_H
+#define FWCORE_FUNCTIONALUTILS_H
+
+#include "Gaudi/Functional/details.h"
+#include "GaudiKernel/DataObjID.h"
+#include "podio/CollectionBase.h"
+
+#include "GaudiKernel/EventContext.h"
+#include "GaudiKernel/ThreadLocalContext.h"
+
+#include <map>
+#include <memory>
+#include <tuple>
+#include <type_traits>
+
+namespace k4FWCore {
+
+  static const std::string frameLocation = "/_Frame";
+
+  namespace details {
+
+    // This function will be used to modify std::shared_ptr<podio::CollectionBase> to the actual collection type
+    template <typename T, typename P> const auto& maybeTransformToEDM4hep(const P& arg) { return arg; }
+    template <typename T, typename P>
+      requires std::is_base_of_v<podio::CollectionBase, P>
+    const auto& maybeTransformToEDM4hep(P* arg) {
+      return *arg;
+    }
+
+    template <typename T, typename P>
+      requires std::same_as<P, std::shared_ptr<podio::CollectionBase>>
+    const auto& maybeTransformToEDM4hep(P arg) {
+      return static_cast<const T&>(*arg);
+    }
+
+    template <typename T, bool addPtr = std::is_base_of_v<podio::CollectionBase, T>>
+    using addPtrIfColl = std::conditional_t<addPtr, std::add_pointer_t<T>, T>;
+
+    template <typename T> const auto& transformtoEDM4hep(const std::shared_ptr<podio::CollectionBase>& arg) {
+      return static_cast<const T>(*arg);
+    }
+
+    // Check if the type is a map like type, where map type is the special map
+    // type to have an arbitrary number of collections as input or output:
+    // std::map<std::string, Coll> where Coll is the collection type
+    template <typename T> struct isMapToCollLike : std::false_type {};
+
+    template <typename Value>
+      requires std::is_base_of_v<podio::CollectionBase, std::remove_cvref_t<Value>>
+    struct isMapToCollLike<std::map<std::string, Value>> : std::true_type {};
```

Comment:
`isMapToCollLike<T>::value` is used quite a few times. Std traits come with helpers to have a slightly less awkward syntax. Here it's not necessary but you might want to provide one
```
template< class T>
inline constexpr bool isMapToCollLike_v= isMapToCollLike<T>::value
```

Reply:
Added `isMapToCollLike_v`